### PR TITLE
Client-Portal Features

### PR DESCRIPTION
**Pull request by minahil3955**
Client Features :
1. Create, view, update, delete Leads
2. Assign Previous/New Client to Lead at Creation
3. Create Phase and Assign it to Existing Lead
4. Create Comments on New lead and phases
5. Navbar created in  Application layout.
6. Showing lists of all Leads, Comments, Phases and Clients
7. Lead promoted to Project on Sale Close
8. Improved Client-Side UI

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -13,3 +13,21 @@
  *= require_tree .
  *= require_self
  */
+a, a:link, a:visited
+{
+  text-decoration: none;
+  color: grey;
+
```
Comment (muhammad-ahmad-shafiq) [inline app/assets/stylesheets/application.scss]: Remove this line.

File: app/assets/stylesheets/application.scss [hunk 2]
```diff
@@ -13,3 +13,21 @@
...
+}
+a:hover {
```
Comment (muhammad-ahmad-shafiq) [inline app/assets/stylesheets/application.scss:21]: Leave a line before a:hover

File: app/assets/stylesheets/application.scss [hunk 3]
```diff
@@ -13,3 +13,21 @@
...
+  color: black;
+}
+body {
```
Comment (muhammad-ahmad-shafiq) [inline app/assets/stylesheets/application.scss:25]: Always leave a line before a new parallel block i.e. body {}

File: app/assets/stylesheets/application.scss [hunk 4]
```diff
@@ -13,3 +13,21 @@
...
+  font-size: 90%;
+}
+button a:link, button a:visited, button a:hover {
+  color:white;
+}
+button {
```
Comment (muhammad-ahmad-shafiq) [inline app/assets/stylesheets/application.scss]: same. leave lines

File: app/controllers/comments_controller.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
+class CommentsController < ApplicationController
+  before_action :authenticate_user!
```
Comment (muhammad-ahmad-shafiq) [inline app/controllers/comments_controller.rb]: You have added this line in most of the controller. There is a better way of doing so.

File: app/controllers/phases_controller.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
+class PhasesController < ApplicationController
```
Comment (muhammad-ahmad-shafiq) [inline app/controllers/phases_controller.rb:1]: I think phase controller route should be nested with the project lead. The route should be like 
/leads/1/phases
/leads/2/phases/new
/leads/3/phases/4
...

Please fix this controller in the above mentioned way.

File: app/controllers/project_leads/comments_controller.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class ProjectLeads::CommentsController < CommentsController
```
Comment (muhammad-ahmad-shafiq) [inline app/controllers/project_leads/comments_controller.rb:1]: 👍

File: app/views/clients/_form.html.erb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+          <p><%= form.label :contact %></p>
+          <%= form.text_field :contact %>
+        </div>
+
+        <div class = 'row'>
+          <div class = 'col-md-8 offset-md-3'>
+            <button action = 'submit'  class='btn btn-dark'> Create Client </button>
+          </div>
+        </div>
+      <% end %>
```
Comment (muhammad-ahmad-shafiq) [inline app/views/clients/_form.html.erb]: Fix the spacing.
It should be like
<div class='asdasdas'>

File: app/views/clients/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        <td><%= link_to 'Edit', edit_client_path(client) %></td>
+        <td><%= link_to 'Delete', client, method: :delete, data: { confirm: 'Are you sure?' } %></td>
+      </tr>
+    <% end %>
+  </tbody>
+</table>
+<br>
+<div class = 'row'>
+  <div class = 'col-md-8 offset-md-5'>
+    <button  class="btn btn-dark mx-5"><%= link_to 'Create New Client', new_client_path %></button>
```
Comment (muhammad-ahmad-shafiq) [inline app/views/clients/index.html.erb:30]: same. spacing issue.

File: app/views/comments/_comments.html.erb [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+<h4><b>Comments<br></h4>
+<% commentable.comments.each do |c| %>
+  <div>
+    <ul>
+      <li>
+        <%= c.body %>
+        <hr>
+      </li>
+    <ul>
+  </div>
```
Comment (muhammad-ahmad-shafiq) [inline app/views/comments/_comments.html.erb]: Why so nested?

File: app/views/comments/_comments.html.erb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+<hr>
+<h4><b>Comments<br></h4>
+<% commentable.comments.each do |c| %>
```
Comment (muhammad-ahmad-shafiq) [inline app/views/comments/_comments.html.erb]: the variable name should be comment not `c` 😢

File: app/views/comments/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+      <th>Creation Date</th>
+      <th colspan="3"></th>
+    </tr>
+  </thead>
+
+  <tbody>
+    <% @comments.each do |comment| %>
+      <tr>
+        <td><%= comment.commentable.name%></td>
+        <td><%=  comment.created_at.day.to_s + '/' + comment.created_at.month.to_s + '/' + comment.created_at.year.to_s %></td>
```
Comment (muhammad-ahmad-shafiq) [inline app/views/comments/index.html.erb]: Use helpers

File: app/views/comments/show.html.erb [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+</p>
+
+<p>
+  <strong>Comment Context:</strong>
+  <%= @comment.body %>
+</p>
+
+<p>
+  <strong>Creation Date:</strong>
+  <td><%=  @comment.created_at.year.to_s + '/' + @comment.created_at.month.to_s + '/' + @comment.created_at.day.to_s %></td>
```
Comment (muhammad-ahmad-shafiq) [inline app/views/comments/show.html.erb]: same. Use helpers

File: app/views/layouts/application.html.erb [hunk 14]
```diff
@@ -11,6 +11,40 @@
...
+      <a class="nav-link" href = '/clients' >Clients</a>
+    </li>
+    <li class="nav-item active">
+    <a class="nav-link" href = '/projects' >Projects</a>
+    </li>
+    <li class="nav-item active">
+      <a class="nav-link" href = '/comments' >Comments</a>
+    </li>
+    </ul>
+    <span class="navbar-text">
```
Comment (muhammad-ahmad-shafiq) [inline app/views/layouts/application.html.erb]: We NEVER use routes or path like that. Always use rails routing helping methods like 
projects_path == '/projects'
project_path(project) == 'projects/<project-id>'

File: app/views/phases/_form.html.erb [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    </div>
+
+    <div class = 'form-group'>
+      <p><%= form.label :due_date %></p>
+      <%= form.date_field :due_date %>
+    </div>
+
+    <div class = 'form-group '>
+      <%= form.label :status %>
+      <%= select_tag :status, options_for_select(Phase.statuses), { :prompt => 'Choose Project Status', :selected => @phase.status } %>
```
Comment (muhammad-ahmad-shafiq) [inline app/views/phases/_form.html.erb]: Use the new way of assing symbol i.e. `propmt: 'Choose Project Stauts'` not `:prompt => 'asdad'`

File: app/views/phases/index.html.erb [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+    </tr>
+  </thead>
+
+  <tbody>
+    <% @phases.each do |phase| %>
+      <tr>
+        <td><%= phase.name %></td>
+        <td><%= phase.due_date %></td>
+        <td><%= phase.status %></td>
+        <td><%=  phase.created_at.year.to_s + '-' + phase.created_at.month.to_s + '-' + phase.created_at.day.to_s %></td>
```
Comment (muhammad-ahmad-shafiq) [inline app/views/phases/index.html.erb]: same

File: config/routes.rb [hunk 17]
```diff
@@ -1,14 +1,26 @@
 Rails.application.routes.draw do
+  resources :phases
+  resources :comments
+  resources :clients
+  resources :project_leads do
+    resources :comments, module: :project_leads
+  end
+  resources :phases do
+    resources :comments, module: :phases
+  end
```
Comment (muhammad-ahmad-shafiq) [inline config/routes.rb]: You have already declared a resource block for `phases` above. You can use that. Why did you write a new one?

File: config/routes.rb [hunk 18]
```diff
@@ -1,14 +1,26 @@
...
+
+  get 'projects', to: 'project_leads#project_index', as: 'projects_index'
```
Comment (muhammad-ahmad-shafiq) [inline config/routes.rb]: There is a better way of doing it

**Review by muhammad-ahmad-shafiq — CHANGES_REQUESTED** [review]

File: app/controllers/comments_controller.rb (hunk 5 above)
Comment (minahil3955) [reply app/controllers/comments_controller.rb]: Add this line in application controller.

File: app/assets/stylesheets/application.scss (hunk 4 above)
Comment (minahil3955) [reply app/assets/stylesheets/application.scss]: Removed

File: app/views/comments/index.html.erb (hunk 12 above)
Comment (minahil3955) [reply app/views/comments/index.html.erb]: i found out but could not get T_T

File: config/routes.rb (hunk 17 above)
Comment (minahil3955) [reply config/routes.rb]: I have made a separate controller for project_lead/comments_controller.rb and phases/comments_controller.rb that's why !

**Pull request merged**